File: src/com/train/algorithm/search/implementInC++/CombinationSumII.cpp
//
// Created by 罗斯 on 5/9/2022.
//
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        sort(candidates.begin(),candidates.end());
        vector<int> cur;
        dfs(candidates,target,0,cur,res);
        return res;
    }
private:
    void dfs(vector<int>& candidates, int target,int s,vector<int>& cur,vector<vector<int>>& res){
        if(target == 0){
            res.push_back(cur);
            return;
        }
        for(int i = s ;i < candidates.size();++i){
            int num = candidates[i];
            if(num > target) return;
            if(i > s && candidates[i] == candidates[i - 1]) continue;
            cur.push_back(num);
            dfs(candidates,target - num,i + 1,cur,res);
            cur.pop_back();
        }
    }
};

File: src/com/train/algorithm/search/implementInPython/CombinationSumII.py
class Solution(object):
    def dfs(self, index, target, arr, res, path):
        if target == 0:
            res.append(path[:])
            return
        for i in range(index, len(arr)):
            if i > index and arr[i] == arr[i-1]:
                continue
            if arr[i] > target:
                break
            path.append(arr[i])
            self.dfs(i + 1, target - arr[i], arr, res, path)
            path.pop()

    def combinationSum2(self, candidates, target):
        candidates.sort()
        res = []
        path = []
        self.dfs(0, target, candidates, res, path)
        return res